src/sciline/visualize.py

```diff
@@ -1,53 +1,97 @@
 # SPDX-License-Identifier: BSD-3-Clause
 # Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
-from typing import Any, Callable, Dict, List, Tuple, get_args, get_origin
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    get_args,
+    get_origin,
+)
 
 from graphviz import Digraph
 
-from .scheduler import Graph
+from .pipeline import Pipeline, SeriesProvider
+from .typing import Graph, Item, Key
 
 
-def to_graphviz(graph: Graph, **kwargs: Any) -> Digraph:
+def to_graphviz(graph: Graph, compact: bool = False, **kwargs: Any) -> Digraph:
     """
     Convert output of :py:class:`sciline.Pipeline.get_graph` to a graphviz graph.
 
     Parameters
     ----------
     graph:
         Output of :py:class:`sciline.Pipeline.get_graph`.
+    compact:
+        If True, parameter-table-dependent branches are collapsed into a single copy
+        of the branch. Recommendend for large graphs with long parameter tables.
     kwargs:
         Keyword arguments passed to :py:class:`graphviz.Digraph`.
     """
     dot = Digraph(strict=True, **kwargs)
-    for p, (p_name, args, ret) in _format_graph(graph).items():
-        dot.node(ret, ret, shape='rectangle')
+    for p, (p_name, args, ret) in _format_graph(graph, compact=compact).items():
+        if '(' in ret and '=' not in ret:
+            dot.node(ret, ret, shape='box3d')
+        else:
+            dot.node(ret, ret, shape='rectangle')
         # Do not draw dummy providers created by Pipeline when setting instances
-        if p_name == 'Pipeline.__setitem__.<locals>.<lambda>':
+        if p_name in (
+            f'{_qualname(Pipeline.__setitem__)}.<locals>.<lambda>',
+            f'{_qualname(Pipeline.set_param_table)}.<locals>.<lambda>',
+        ):
             continue
-        dot.node(p, p_name, shape='ellipse')
-        for arg in args:
-            dot.node(arg, arg, shape='rectangle')
-            dot.edge(arg, p)
-        dot.edge(p, ret)
+        # Do not draw the internal provider gathering index-dependent results into
+        # a dict
+        if p_name == _qualname(SeriesProvider):
+            for arg in args:
+                dot.edge(arg, ret, style='dashed')
+        else:
+            dot.node(p, p_name, shape='ellipse')
+            for arg in args:
+                dot.edge(arg, p)
+            dot.edge(p, ret)
     return dot
 
 
-def _format_graph(graph: Graph) -> Dict[str, Tuple[str, List[str], str]]:
+def _qualname(obj: Any) -> Any:
+    return (
+        obj.__qualname__ if hasattr(obj, '__qualname__') else obj.__class__.__qualname__
+    )
```

Comment:
```suggestion
def _qualname(obj: Any) -> Any:
    return getattr(obj, '__qualname__', obj.__class__.__qualname__)
```

Reply:
Won't that raise if `obj` has no `__class__`?

Reply:
True. Leave it as is